File: solutions/_560_subarray_sum_equals_k/solution.java
package _560_subarray_sum_equals_k;

import java.util.*;

class Solution {

  public int subarraySum(int[] nums, int k) {
    int count = 0;
    int sum = 0;

    Map<Integer, Integer> map = new HashMap<>();
    map.put(0, 1);

    for (int i = 0; i < nums.length; i++) {
      sum += nums[i];
      int complement = sum - k;
      if (map.containsKey(complement)) {
        count += map.get(complement);
      }
      map.put(sum, map.getOrDefault(sum, 0) + 1);
    }

    return count;
  }
}
/*
https://leetcode.com/problems/subarray-sum-equals-k/
*/

File: solutions/_560_subarray_sum_equals_k/solution.py
from typing import List


class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        count = 0
        sum = 0
        dic = {}
        dic[0] = 1
        for i in range(len(nums)):
            sum += nums[i]
            complement = sum - k
            if complement in dic:
                count += dic[complement]
            dic[sum] = dic.get(sum, 0) + 1
        return count


# https://leetcode.com/problems/subarray-sum-equals-k/
